hw3/02-charts.js

```diff
@@ -24,22 +64,89 @@ const borderColors = [
   'rgba(40, 159, 64, 1)',
   'rgba(210, 199, 199, 1)',
   'rgba(78, 52, 199, 1)',
+  ...moreColors,
 ];
 
+function capitalizeHouseNames(str) {
+  return str
+    .split(' ')
+    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
+    .join(' ');
+}
+
+function standardizeHouseName(name) {
+  if (!name) return name;
+
+  name = name.trim().toLowerCase();
+
+  if (name.includes('stark')) {
+    return 'House Stark';
+  } else if (name.includes('lannister') || name.includes('lanister')) {
+    return 'House Lannister';
+  } else if (
+    name.includes('targaryen') ||
+    name.includes('targaryn') ||
+    name.includes('targaryan')
+  ) {
+    return 'House Targaryen';
+  } else if (name.includes('baratheon')) {
+    return 'House Baratheon';
+  } else if (name.includes('greyjoy')) {
+    return 'House Greyjoy';
+  } else if (name.includes('martell')) {
+    return 'House Martell';
+  } else if (name.includes('arryn')) {
+    return 'House Arryn';
+  } else if (name.includes('frey')) {
+    return 'House Frey';
+  } else if (name.includes('tyrell')) {
+    return 'House Tyrell';
+  } else if (name.includes('tully')) {
+    return 'House Tully';
+  } else if (
+    name.includes('none') ||
+    name.includes('unknown') ||
+    name.includes('unkown')
+  ) {
+    return 'House Unknown';
+  } else {
+    // For all other names, capitalize the name given. As it is their last name most of the time
+    // aka they are their own house. And other houses that are correct just caps em
+    return capitalizeHouseNames(name);
+  }
+}
+
 // url for the Thrones API
 const url = 'https://thronesapi.com/api/v2/Characters';
 
-const renderChart = () => {
-  const donutChart = document.querySelector('.donut-chart');
+const renderChart = async () => {
+  const fetchApiData = async () => {
+    const response = await fetch(url);
+    const data = await response.json();
+    return data;
+  };
+  const data = await fetchApiData();
+  console.log(data);
+  const familyCounts = data.reduce((acc, char) => {
+    const correctedFamilyName = standardizeHouseName(char.family); // fix the names if need be
+    if (char.family) {
+      acc[correctedFamilyName] = (acc[correctedFamilyName] || 0) + 1; // map by family name {houseName: count}
+    }
+    return acc;
+  }, {});
+
+  const labels = Object.keys(familyCounts); // get keys aka houseNames
+  const familyData = Object.values(familyCounts); // get the values aka count
+  const donutChart = document.querySelector('.donut-chart'); // find where to append it
 
   new Chart(donutChart, {
```

Comment:
The image provided in the homework technically doesn't have a legend in it, but it also doesn't specify that it shouldn't have one.

Reply:
hmm, true. I guess I just left it in. I like it. I remeber that you removed it, so I suppose that yours does match more closely